gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java

```diff
@@ -0,0 +1,274 @@
+   *
+   * @see BlobWriteSessionConfigs#bufferToTempDirThenUpload()
+   */
+  private static final String DEFAULT_TEMP_SUB_DIR = "google-cloud-storage";
+
+  private static GoogleCloudStorage helperGcs;
+
+  private static ImmutableSet<String> tempDirs =
+      ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
+  private static ImmutableSet<Path> tempDirsPath =
+      tempDirs.stream().map(x -> Paths.get(x)).collect(ImmutableSet.toImmutableSet());
+
+  @Rule
+  public TestName name =
+      new TestName() {
+        // With parametrization method name will get [index] appended in their name.
+        @Override
+        public String getMethodName() {
+          return super.getMethodName().replaceAll("[\\[,\\]]", "");
+        }
+      };
+
+  @BeforeClass
+  public static void before() throws IOException {
+
+    helperGcs = GoogleCloudStorageTestHelper.createGcsClientImpl();
+    helperGcs.createBucket(TEST_BUCKET);
+  }
+
+  @AfterClass
+  public static void after() throws IOException {
+    try {
+      BUCKET_HELPER.cleanup(helperGcs);
+    } finally {
+      helperGcs.close();
+    }
+  }
+
+  @Before
+  public void setUp() {
+    System.setProperty("java.io.tmpdir", GCS_WRITE_TMP_DIR);
+  }
+
+  @After
+  public void cleanUp() {
+    ImmutableSet<String> tempDirs = ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
+    Iterator<String> iterator = tempDirs.stream().iterator();
+    while (iterator.hasNext()) {
+      String filePath = iterator.next();
+      File directory = new File(filePath);
+      if (directory.listFiles() != null) {
+        for (File file : new File(filePath).listFiles()) {
+          file.delete();
+        }
+      }
+    }
+  }
+
+  @Test
+  public void writeToDiskDisabled() throws IOException {
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                AsyncWriteChannelOptions.builder().setUploadType(UploadType.DEFAULT).build())
+            .build();
+
+    GoogleCloudStorage gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    byte[] bytesToWrite = new byte[1024];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    Path temporaryPath = Paths.get(GCS_WRITE_TMP_DIR, DEFAULT_TEMP_SUB_DIR);
+    // validate that there were no files
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 0);
+
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // no temporary files in
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 0);
+
+    writeChannel.close();
+    // temporary files will be deleted from disk once upload is finished.
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 0);
+
+    gcs.close();
+  }
+
+  @Test
+  public void writeToDefaultPathThenUploadEnabled() throws IOException {
```

Comment:
Lot of code duplication between test. Please refactor and reuse.

Reply:
There are three things a test is doing
1. Setting up storage option ( different fro different test)
2. Creating gcsImpl from option (similar for all test, already refactored)
3. Writing data to file (common for all test, not refactored)
4. Verification (different for all test, refactored with common functions)

The only remaining scope of refactoring is step(3) which is not huge and don't see scope for any further meaningful refactoring.

Reply:
All three tests are similar and has code duplication

Reply:
tied another way.